In [311]:
import numpy as np
import pandas as pd
from scipy.stats import chi2_contingency
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline
import math
import numpy

In [312]:
df=pd.read_csv('StudentsPerformance.csv')
df

gender race/ethnicity parental level of education         lunch  \
0    female        group C            some high school  free/reduced   
1    female        group B                 high school  free/reduced   
2    female        group B            some high school  free/reduced   
3    female        group B                some college      standard   
4    female        group C                some college  free/reduced   
..      ...            ...                         ...           ...   
995    male        group E                some college      standard   
996    male        group A                some college      standard   
997    male        group D                some college      standard   
998    male        group E          associate's degree  free/reduced   
999    male        group E           bachelor's degree      standard   

    test preparation course  math score  reading score  writing score  
0                      none           0             17             10  
1                      none           8             24             23  
2                      none          18             32             28  
3                      none          11             38             32  
4                      none          22             39             33  
..                      ...         ...            ...            ...  
995               completed          99             87             81  
996               completed         100             96             86  
997               completed         100             97             99  
998               completed         100            100             93  
999               completed         100            100            100  

[1000 rows x 8 columns]

In [313]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 8 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   gender                       1000 non-null   object
 1   race/ethnicity               1000 non-null   object
 2   parental level of education  1000 non-null   object
 3   lunch                        1000 non-null   object
 4   test preparation course      1000 non-null   object
 5   math score                   1000 non-null   int64 
 6   reading score                1000 non-null   int64 
 7   writing score                1000 non-null   int64 
dtypes: int64(3), object(5)
memory usage: 62.6+ KB


## 1. Find out how many males and females participated in the test.

In [314]:
df['gender'].value_counts()

female    518
male      482
Name: gender, dtype: int64

.No of males and females participated in the test are 482 males and 518 females

In [315]:
df.groupby('gender')['test preparation course'].value_counts()


gender  test preparation course
female  none                       334
        completed                  184
male    none                       308
        completed                  174
Name: test preparation course, dtype: int64

.From the 518 females only 184 completed the test preparation course and 334 coudn't complete
From the total of 482 males only 174 completed the test preparation course and 308 couldn't complete

# 2. What do you think about the students' parental level of education?

In [316]:
df.groupby('gender')['parental level of education'].value_counts()


gender  parental level of education
female  some college                   118
        associate's degree             116
        high school                     94
        some high school                91
        bachelor's degree               63
        master's degree                 36
male    some college                   108
        associate's degree             106
        high school                    102
        some high school                88
        bachelor's degree               55
        master's degree                 23
Name: parental level of education, dtype: int64

.parental level education in males and females

In [317]:
df['parental level of education'].value_counts()

some college          226
associate's degree    222
high school           196
some high school      179
bachelor's degree     118
master's degree        59
Name: parental level of education, dtype: int64

parental level of education as total

## 3. Who scores the most on average for math, reading and writing based on
● Gender
● Test preparation course

In [318]:
df.groupby('gender').mean()

math score  reading score  writing score
gender                                          
female   63.376448      72.590734      72.467181
male     68.821577      65.545643      63.446058

Female students scored most for reading and writing .Male students sored most in maths

In [319]:
df.groupby('test preparation course').mean()

math score  reading score  writing score
test preparation course                                          
completed                  69.96648      74.175978      74.684358
none                       63.78972      66.417445      64.457944

.Test preparation course completed students score most than those who have not completed

In [320]:
df.groupby(['gender','test preparation course']).mean()

math score  reading score  writing score
gender test preparation course                                          
female completed                 67.108696      77.375000      78.793478
       none                      61.320359      69.955090      68.982036
male   completed                 72.988506      70.793103      70.339080
       none                      66.467532      62.581169      59.551948

## 4. What do you think about the scoring variation for math, reading and writing
based on
● Gender
● Test preparation course

In [321]:
df.groupby('gender').var()

math score  reading score  writing score
gender                                          
female  256.958593     207.677438     220.369327
male    211.889097     200.211010     202.413924

when comparing the variation in maths,reading and writing score female has more variation than males

In [322]:
df.groupby('test preparation course').var()

math score  reading score  writing score
test preparation course                                          
completed                210.884027     183.265864     175.202612
none                     246.668662     213.419851     226.251739

. when comparing the variationon test preparation course completed is less than not completed ones

In [323]:
df.groupby(['gender','test preparation course']).var()

math score  reading score  writing score
gender test preparation course                                          
female completed                208.173913     161.175546     149.836897
       none                     272.602767     214.241220     225.495172
male   completed                197.098133     185.297987     166.098233
       none                     205.513600     184.843553     181.394687

.we took both the parameters together .Results are same but there is slight difference in number

## 5. The management needs your help to give bonus points to the top 25% of
students based on their math score, so how will you help the management
to achieve this.

In [324]:
df1 = df.sort_values(by=['math score'],ascending=False)
df1

gender race/ethnicity parental level of education         lunch  \
999    male        group E           bachelor's degree      standard   
996    male        group A                some college      standard   
515  female        group E                some college      standard   
517  female        group E          associate's degree      standard   
516  female        group E           bachelor's degree      standard   
..      ...            ...                         ...           ...   
2    female        group B            some high school  free/reduced   
51   female        group C            some high school  free/reduced   
3    female        group B                some college      standard   
1    female        group B                 high school  free/reduced   
0    female        group C            some high school  free/reduced   

    test preparation course  math score  reading score  writing score  
999               completed         100            100            100  
996               completed         100             96             86  
515                    none         100             92             97  
517                    none         100            100            100  
516                    none         100            100            100  
..                      ...         ...            ...            ...  
2                      none          18             32             28  
51                     none          13             41             51  
3                      none          11             38             32  
1                      none           8             24             23  
0                      none           0             17             10  

[1000 rows x 8 columns]

In [325]:
d=df1.quantile(0.75)
d.head()


math score       77.00
reading score    79.25
writing score    79.00
Name: 0.75, dtype: float64

In [328]:
df1['math score'] >=77

999     True
996     True
515     True
517     True
516     True
       ...  
2      False
51     False
3      False
1      False
0      False
Name: math score, Length: 1000, dtype: bool

In [336]:
df1[df1['math score']>=d['math score']]
#df1[df1['math score']>=77 *alternative code

gender race/ethnicity parental level of education         lunch  \
999    male        group E           bachelor's degree      standard   
996    male        group A                some college      standard   
515  female        group E                some college      standard   
517  female        group E          associate's degree      standard   
516  female        group E           bachelor's degree      standard   
..      ...            ...                         ...           ...   
851    male        group D                some college      standard   
849    male        group A           bachelor's degree      standard   
850    male        group C          associate's degree  free/reduced   
421  female        group C                some college  free/reduced   
414  female        group A                 high school  free/reduced   

    test preparation course  math score  reading score  writing score  
999               completed         100            100            100  
996               completed         100             96             86  
515                    none         100             92             97  
517                    none         100            100            100  
516                    none         100            100            100  
..                      ...         ...            ...            ...  
851               completed          77             62             62  
849                    none          77             67             68  
850                    none          77             67             64  
421                    none          77             90             91  
414               completed          77             88             85  

[255 rows x 8 columns]

.shows the index of the students who scored top 25% in math score

In [335]:
df1.loc[(df1['math score']) >=(d['math score'])]

gender race/ethnicity parental level of education         lunch  \
999    male        group E           bachelor's degree      standard   
996    male        group A                some college      standard   
515  female        group E                some college      standard   
517  female        group E          associate's degree      standard   
516  female        group E           bachelor's degree      standard   
..      ...            ...                         ...           ...   
851    male        group D                some college      standard   
849    male        group A           bachelor's degree      standard   
850    male        group C          associate's degree  free/reduced   
421  female        group C                some college  free/reduced   
414  female        group A                 high school  free/reduced   

    test preparation course  math score  reading score  writing score  
999               completed         100            100            100  
996               completed         100             96             86  
515                    none         100             92             97  
517                    none         100            100            100  
516                    none         100            100            100  
..                      ...         ...            ...            ...  
851               completed          77             62             62  
849                    none          77             67             68  
850                    none          77             67             64  
421                    none          77             90             91  
414               completed          77             88             85  

[255 rows x 8 columns]

In [338]:
df.nlargest(250,'math score')
#df1.nlargest(250,'math score') * alternative code

gender race/ethnicity parental level of education         lunch  \
515  female        group E                some college      standard   
516  female        group E           bachelor's degree      standard   
517  female        group E          associate's degree      standard   
996    male        group A                some college      standard   
997    male        group D                some college      standard   
..      ...            ...                         ...           ...   
428  female        group D                some college  free/reduced   
429  female        group E            some high school      standard   
849    male        group A           bachelor's degree      standard   
850    male        group C          associate's degree  free/reduced   
851    male        group D                some college      standard   

    test preparation course  math score  reading score  writing score  
515                    none         100             92             97  
516                    none         100            100            100  
517                    none         100            100            100  
996               completed         100             96             86  
997               completed         100             97             99  
..                      ...         ...            ...            ...  
428                    none          77             86             86  
429                    none          77             79             80  
849                    none          77             67             68  
850                    none          77             67             64  
851               completed          77             62             62  

[250 rows x 8 columns]